### PR TITLE
Helix fail if exceptions thrown during test discovery

### PR DESCRIPTION
**Pull request by HaoK**

### CODE REVIEW

**Comment by anurse** [comment]
Looks like a git checkout problem? Cancelling and restarting.

**Comment by HaoK** [comment]
:( of course the PR that I'm trying to make red is green...

**Comment by HaoK** [comment]
@BrennanConroy looks like the StackExchangeRedis.Tests are working better now, do you want to try on this branch for https://github.com/aspnet/AspNetCore/issues/7746 ?

**Comment by BrennanConroy** [comment]
I fixed the root cause of the SignalR issue with https://github.com/aspnet/AspNetCore/commit/337370263a7119ee5b8e0a6649d9e2da2fd9f0c3
So you'll need to either remove the fix in this PR to test it or do something else.

**Comment by HaoK** [comment]
Cool I'll revert your fix to test the fix

File: src/SignalR/Directory.Build.props [hunk 1]
```diff
@@ -16,8 +16,8 @@
     <Content Include="$(MSBuildThisFileDirectory)xunit.runner.json" Link="xunit.runner.json">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </Content>
-    <Reference Include="Microsoft.Extensions.Logging.Testing" />
-  </ItemGroup>
+      <Reference Include="Microsoft.Extensions.Logging.Testing" PrivateAssets="All" />
```
Comment (HaoK) [inline src/SignalR/Directory.Build.props]: Revert this before merging

**Comment by HaoK** [comment]
Discovery exceptions will now fail the tests, see https://mc.dot.net/#/user/aspnetcore/pr~2Faspnet~2Faspnetcore/ci/20190304.1

File: eng/helix/vstest/runtests.sh [hunk 2]
```diff
@@ -60,4 +60,11 @@ export DOTNET_CLI_HOME="$HELIX_CORRELATION_PAYLOAD/home"
...
 
-$HELIX_CORRELATION_PAYLOAD/sdk/dotnet vstest $1 --logger:trx
+$DOTNET_ROOT/dotnet vstest $1 -lt >discovered.txt
+if grep -q "Exception thrown" discovered.txt; then
+    echo -e "${RED}Exception thrown during test discovery${RESET}".
+    cat discovered.txt
+    exit 1
+fi
+
+$DOTNET_ROOT/dotnet vstest $1 --logger:trx
```
Comment (dougbu) [inline eng/helix/vstest/runtests.sh]: Another inconsistency here is the lack of `--logger:console;verbosity=normal`. Is that intentional?

**Review by dougbu — APPROVED** [review]
To confirm: This added logging will help find the root cause i.e. there's more work to do here?

**Comment by pakrym** [comment]
I think we need to follow up on this with vstest/xunit

**Comment by HaoK** [comment]
@dougbu there's no additional work for this at this time, it just fixes helix treating test projects that have exceptions during test discovery from silently passing

@pakrym I think this behavior is by design since the error code from xunit is equal to the number of tests that fail. We can maybe ask them to return -1 or something

File: eng/helix/vstest/runtests.sh (hunk 2 above)
Comment (HaoK) [reply eng/helix/vstest/runtests.sh]: @pakrym added these, most likely for windows specific IIS things, should we just add them to be consistent?

Comment (HaoK) [reply eng/helix/vstest/runtests.sh]: Or maybe we should switch this to a HelixLogSettings so test projects can customize on demand and feed it into the scripts, not sure this is something that we need more granularity on...

Comment (dougbu) [reply eng/helix/vstest/runtests.sh]: Suggest making it consistent. If we have trouble with excess logging, can revisit in the future.

Comment (pakrym) [reply eng/helix/vstest/runtests.sh]: I'm making it consistent in https://github.com/aspnet/AspNetCore/pull/8094

**Comment by pakrym** [comment]
> I think this behavior is by design since the error code from xunit is equal to the number of tests that fail. We can maybe ask them to return -1 or something

We should change it, not running any tests shouldn't exit with the successfull error code.

File: eng/helix/vstest/runtests.sh (hunk 2 above)
Comment (HaoK) [reply eng/helix/vstest/runtests.sh]: Ok I'll leave this for now in this PR if you are changing it, I'm just rerunning helix before merging, there was a flaky helix failure

**Comment by HaoK** [comment]
Where should I file a bug for the vstest xunit not failing issue?

**Comment by pakrym** [comment]
There is https://github.com/Microsoft/vstest/issues/1899. If the claim is that runner should fail http://github.com/xunit/xunit/

**Comment by HaoK** [comment]
Ok yeah https://github.com/Microsoft/vstest/issues/1899 is basically the issue...

Looks like there's something flaky in InMemory.FunctionalTests still, as the test hung and was killed
https://mc.dot.net/#/user/aspnetcore/pr~2Faspnet~2Faspnetcore/ci/20190304.3/workItem/InMemory.FunctionalTests-netcoreapp3.0/wilogs

@BrennanConroy also a weird discovery test failure in signal r too but only on ubuntu, I'll rerun again since it didn't happen on the previous run:

https://mc.dot.net/#/user/aspnetcore/pr~2Faspnet~2Faspnetcore/ci/20190304.3/workItem/InMemory.FunctionalTests-netcoreapp3.0/wilogs

```
Exception thrown during test discovery.
Microsoft (R) Test Execution Command Line Tool Version 16.0.0-preview-20181205-02
Copyright (c) Microsoft Corporation.  All rights reserved.

The following Tests are available:
[xUnit.net 00:00:05.98] Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests: Exception thrown during theory discovery on 'Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisEndToEndTests.HubConnectionCanSendAndReceiveMessages'; falling back to single test case.
System.InvalidOperationException: No process is associated with this object.
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.GetWaitState()
   at System.Diagnostics.Process.WaitForExitCore(Int32 milliseconds)
   at System.Diagnostics.Process.WaitForExit()
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.RunProcessAndWait(String fileName, String arguments, String prefix, ILogger logger, TimeSpan timeout, String& output) in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 163
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.RunCommand(String commandAndArguments, String prefix, ILogger logger, String& output) in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 135
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.RunCommand(String commandAndArguments, String prefix, String& output) in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 131
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.Create() in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 44
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.get_Default() in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 23
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.SkipIfDockerNotPresentAttribute.CheckDocker() in /_/src/SignalR/server/StackExchangeRedis/test/SkipIfDockerNotPresentAttribute.cs:line 17
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.SkipIfDockerNotPresentAttribute.get_IsMet() in /_/src/SignalR/server/StackExchangeRedis/test/SkipIfDockerNotPresentAttribute.cs:line 12
   at Microsoft.AspNetCore.Testing.xunit.TestMethodExtensions.EvaluateSkipConditions(ITestMethod testMethod)
   at Microsoft.Extensions.Logging.Testing.LoggedConditionalTheoryDiscoverer.CreateTestCasesForDataRow(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo theoryAttribute, Object[] dataRow)
   at Xunit.Sdk.TheoryDiscoverer.Discover(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo theoryAttribute) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TheoryDiscoverer.cs:line 238
[xUnit.net 00:00:05.99] Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests: Exception during discovery:
System.InvalidOperationException: No process is associated with this object.
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.GetWaitState()
   at System.Diagnostics.Process.WaitForExitCore(Int32 milliseconds)
   at System.Diagnostics.Process.WaitForExit()
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.RunProcessAndWait(String fileName, String arguments, String prefix, ILogger logger, TimeSpan timeout, String& output) in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 163
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.RunCommand(String commandAndArguments, String prefix, ILogger logger, String& output) in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 135
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.RunCommand(String commandAndArguments, String prefix, String& output) in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 131
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.Create() in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 44
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
--- End of stack trace from previous location where exception was thrown ---
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.Docker.get_Default() in /_/src/SignalR/server/StackExchangeRedis/test/Docker.cs:line 23
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.SkipIfDockerNotPresentAttribute.CheckDocker() in /_/src/SignalR/server/StackExchangeRedis/test/SkipIfDockerNotPresentAttribute.cs:line 17
   at Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.SkipIfDockerNotPresentAttribute.get_IsMet() in /_/src/SignalR/server/StackExchangeRedis/test/SkipIfDockerNotPresentAttribute.cs:line 12
   at Microsoft.AspNetCore.Testing.xunit.TestMethodExtensions.EvaluateSkipConditions(ITestMethod testMethod)
   at Microsoft.Extensions.Logging.Testing.LoggedConditionalTheoryDiscoverer.CreateTestCasesForTheory(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo theoryAttribute)
   at Xunit.Sdk.TheoryDiscoverer.Discover(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo theoryAttribute) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TheoryDiscoverer.cs:line 263
   at Microsoft.Extensions.Logging.Testing.LoggedTestFrameworkDiscoverer.FindTestsForMethod(ITestMethod testMethod, Boolean includeSourceInformation, IMessageBus messageBus, ITestFrameworkDiscoveryOptions discoveryOptions)
   at Xunit.Sdk.XunitTestFrameworkDiscoverer.FindTestsForType(ITestClass testClass, Boolean includeSourceInformation, IMessageBus messageBus, ITestFrameworkDiscoveryOptions discoveryOptions) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\XunitTestFrameworkDiscoverer.cs:line 132
   at Xunit.Sdk.TestFrameworkDiscoverer.FindTestsForTypeAndWrapExceptions(ITestClass testClass, Boolean includeSourceInformation, IMessageBus messageBus, ITestFrameworkDiscoveryOptions discoveryOptions) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestFrameworkDiscoverer.cs:line 156
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisDependencyInjectionExtensionsTests.AddRedisWithConnectionStringProperlyParsesOptions(connectionString: "testredis.example.com", host: "testredis.example.com", port: 0, password: null, useSsl: False)
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisDependencyInjectionExtensionsTests.AddRedisWithConnectionStringProperlyParsesOptions(connectionString: "testredis.example.com:6380,ssl=True", host: "testredis.example.com", port: 6380, password: null, useSsl: True)
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisDependencyInjectionExtensionsTests.AddRedisWithConnectionStringProperlyParsesOptions(connectionString: "testredis.example.com:6380,password=hunter2,ssl=Tr"..., host: "testredis.example.com", port: 6380, password: "hunter2", useSsl: True)
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisHubLifetimeManagerTests.CamelCasedJsonIsPreservedAcrossRedisBoundary
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseAck(testName: "Zero")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseAck(testName: "Fixnum")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseAck(testName: "Uint8")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseAck(testName: "Uint16")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseAck(testName: "Uint32")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteAck(testName: "Zero")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteAck(testName: "Fixnum")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteAck(testName: "Uint8")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteAck(testName: "Uint16")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteAck(testName: "Uint32")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseGroupCommand(testName: "GroupAdd")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseGroupCommand(testName: "GroupRemove")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteGroupCommand(testName: "GroupAdd")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteGroupCommand(testName: "GroupRemove")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseInvocation(testName: "NoExcludedIds")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseInvocation(testName: "OneExcludedId")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.ParseInvocation(testName: "ManyExcludedIds")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteInvocation(testName: "NoExcludedIds")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteInvocation(testName: "OneExcludedId")
    Microsoft.AspNetCore.SignalR.StackExchangeRedis.Tests.RedisProtocolTests.WriteInvocation(testName: "ManyExcludedIds")
```

**Comment by BrennanConroy** [comment]
Yeah I'm aware of that one. I'll probably have a change sometime today for it

**Comment by anurse** [comment]
> Looks like there's something flaky in InMemory.FunctionalTests still, as the test hung and was killed

Probably https://github.com/aspnet/AspNetCore/issues/8164

**Comment by HaoK** [comment]
Ok since those helix failures are understood I'll just merge this now

**Comment by pakrym** [comment]
@BrennanConroy please address the discovery failure ASAP

@HaoK please avoid merging changes that leave build permanently red even if they reveal existing issues.

cc @anurse

**Comment by HaoK** [comment]
Sorry it wasn't clear to me that its wasn't just a flaky signal r test since it passed the previous run.

**Pull request merged**